runtime/tchannel_client.go

```diff
@@ -43,150 +47,352 @@ type tchannelClient struct {
-	if err := WriteHeaders(twriter, headers); err != nil {
-		_ = twriter.Close()
+	// write logs
+	c.logger.Info("Finished an outgoing client TChannel request", c.logFields()...)
+}
 
-		return errors.Wrapf(err, "could not write headers for outbound call %s: ", c.serviceName)
+func (c *tchannelCall) logFields() []zapcore.Field {
+	fields := []zapcore.Field{
+		zap.String("remoteAddr", c.call.RemotePeer().HostPort),
+		zap.Time("timestamp-started", c.startTime),
+		zap.Time("timestamp-finished", c.finishTime),
+
+		// TODO: Do not log body by default because PII and bandwidth.
+		// Temporarily log during the development cycle
+		// TODO: Add a gateway level configurable body unmarshaller
+		// to extract only non-PII info.
+		zap.ByteString("Request Body", c.reqBody),
+		zap.ByteString("Response Body", c.resBody),
+	}
+
+	for k, v := range c.reqHeaders {
+		fields = append(fields, zap.String("Request-Header-"+k, v))
+	}
+	for k, v := range c.resHeaders {
+		fields = append(fields, zap.String("Response-Header-"+k, v))
+	}
+
+	// TODO: log jaeger trace span
+
+	return fields
+}
+
+// writeReqHeaders writes request headers to arg2
+func (c *tchannelCall) writeReqHeaders(reqHeaders map[string]string) error {
+	c.reqHeaders = reqHeaders
+
+	twriter, err := c.call.Arg2Writer()
+	if err != nil {
+		c.logger.Error("Could not create arg2writer for outbound request", zap.Error(err))
+		return errors.Wrapf(
+			err, "Could not create arg2writer for outbound %s.%s (%s %s) request",
+			c.client.clientID, c.methodName, c.client.serviceName, c.serviceMethod,
+		)
+	}
+	if err := WriteHeaders(twriter, reqHeaders); err != nil {
+		_ = twriter.Close()
+		c.logger.Error("Could not write headers for outbound request", zap.Error(err))
+		return errors.Wrapf(
+			err, "Could not write headers for outbound %s.%s (%s %s) request",
+			c.client.clientID, c.methodName, c.client.serviceName, c.serviceMethod,
+		)
 	}
 	if err := twriter.Close(); err != nil {
-		return errors.Wrapf(err, "could not close arg2writer for outbound call %s: ", c.serviceName)
+		c.logger.Error("Could not close arg2writer for outbound request", zap.Error(err))
+		return errors.Wrapf(
+			err, "Could not close arg2writer for outbound %s.%s (%s %s) request",
+			c.client.clientID, c.methodName, c.client.serviceName, c.serviceMethod,
+		)
 	}
 
-	twriter, err = call.Arg3Writer()
+	return nil
+}
+
+// writeReqBody writes request body to arg3
+func (c *tchannelCall) writeReqBody(req RWTStruct) error {
+	structWireValue, err := req.ToWire()
 	if err != nil {
-		return errors.Wrapf(err, "could not create arg3writer for outbound call %s: ", c.serviceName)
+		c.logger.Error("Could not write request for outbound request", zap.Error(err))
+		return errors.Wrapf(
+			err, "Could not write request for outbound %s.%s (%s %s) request",
+			c.client.clientID, c.methodName, c.client.serviceName, c.serviceMethod,
+		)
 	}
+	c.reqBody = structWireValue.GetBinary()
```

Comment:
`GetBinary()` does not do what you think it does. This `reqBody` is not useful here.

Reply:
Removing `reqBody`.